narrow-state: Clarify channel operand functions and documentation for current filter.

laurynmm opened the pull request:
Prep commits for work towards [#32959](https://github.com/zulip/zulip/issues/32959):
- Updates comments in `narrow_state.ts` for the fact that we now expect a operand string for a canonical "channel" operator to be the channel ID or an empty string for a current `Filter` object.
- Updates `narrow_state.stream_sub` to use `narrow_state.stream_id`:
  - De-duplicates the shared logic between these two functions.
  - Catches a subtle bug with how the `narrow_state` data was set up for a test in `buddy_data.test.cjs`.
- Explicitly note and check that `narrow_state.stream_id` can return "NaN" in the case where the channel operand is an empty string.
- Updates two functions in `buddy_data.ts.` to use the updated `narrow_state.stream_sub` function when getting or checked subscriber data as this will return data for a tracked/known channel and "undefined" for an untracked/unknown channel.

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/narrow_state.ts:
```diff
@@ -222,7 +222,7 @@ export let get_first_unread_info = (
+        // We're in either the inbox or recent conversations view.
         blueslip.error("unexpected call to get_first_unread_info");
         return cannot_compute_response;
     }
```
This function is always called from a `MessageListData.filter` when entering a message feed view, which is definitely an actual message list. I think this block should be deleted, and the function declared as taking a `Filter` object passed by the caller always. This is cleanup that @amanagr and I missed after reworking `home_msg_list` to use a `Filter` object.

laurynmm replied:
Added a prep commit for this refactor of `get_first_unread_info` to take a declared `Filter` parameter.

timabbott commented on web/src/narrow_state.ts:
```diff
@@ -398,14 +398,9 @@ export function narrowed_by_stream_reply(current_filter: Filter | undefined = fi
         return false;
     }
-
     return stream_id === narrow_sub.stream_id;
```
This is probably cleaner calling `narrow_state.stream_id`, since we don't actually need the `sub` object, just the ID.

laurynmm replied:
Made the update so that this calls `narrow_state.stream_id` as we don't need to worry about whether this is a known channel ID or not for the callers of this function.

timabbott commented on web/src/buddy_data.ts:
```diff
@@ -347,7 +347,7 @@ function maybe_shrink_list(
     // We want to always show PM recipients even if they're inactive.
     const pm_ids_set = narrow_state.pm_ids_set();
-    const stream_id = narrow_state.stream_id();
+    const stream_id = narrow_state.stream_sub()?.stream_id;
```
This looks correct, but I think someone will feel tempted to refactor this back to the original code. I'm not sure if there's a good way to define an abstraction, like `narrow_state.have_metadata_stream_id`, for this?

@evykassirer should probably review this commit.

evykassirer replied:
Can we instead have an optional param passed to `narrow_state.stream_id` like `only_valid_id` or something? I think it would be nicer to make it really clear what `stream_id` returns instead of sometimes calling one of these vs the other which look very similar.

laurynmm replied:
Okay. I went with @evykassirer's suggestion to add a boolean parameter to `narrow_state.stream_id` that defaults to `false` as that's the current behavior.

timabbott commented on web/src/narrow_state.ts:
```diff
@@ -131,6 +131,8 @@ export let stream_id = (current_filter: Filter | undefined = filter()): number |
     if (stream_operands.length === 1 && stream_operands[0] !== undefined) {
+        // If the operand is not a valid channel ID, then
+        // it is an empty string, and this will return NaN.
         return Number.parseInt(stream_operands[0], 10);
```
I think it'd be better to only have one class of invalid value coming out of this function -- let's just have this do the `NaN` check and replace it with returning `undefined`.

laurynmm replied:
Done. Added that as a separate commit in these changes.

evykassirer commented on web/src/narrow_state.ts:
```diff
@@ -398,14 +398,9 @@ export function narrowed_by_stream_reply(current_filter: Filter | undefined = fi
 export function is_for_stream_id(stream_id: number, filter?: Filter): boolean {
-    // This is not perfect, since we still track narrows by
-    // name, not id, but at least the interface is good going
-    // forward.
```
Nice! Love seeing deletions of comments like this :)

evykassirer commented on web/tests/buddy_data.test.cjs:
```diff
@@ -424,7 +425,7 @@ test("get_conversation_participants", ({override_rewire}) => {
     const rome_sub = {name: "Rome", subscribed: true, stream_id: 1001};
     stream_data.add_sub(rome_sub);
-    peer_data.set_subscribers(rome_sub.stream_id, [selma.user_id]);
+    peer_data.set_subscribers(rome_sub.stream_id, [selma.user_id, me.user_id]);
```
I'm curious if this was needed to stop a test from failing after you made these changes (and if so, why)! Or if it just made more sense to have `me` as a subscriber because they're subscribed to this channel.

laurynmm replied:
No, this change was not to stop a test from failing. It was just me noticing that the `rome_sub` object had them as `subscribed: true`. And if that's true, then we'd expect the peer_data to have the current user as a subscriber. I can break this into a separate commit to make that clearer.

evykassirer replied:
Yeah I think a separate commit makes more sense to me!

timabbott replied:
Split a separate commit for this before merging.

laurynmm replied:
Thanks! I will note that the other test in this commit did fix a failing test with the changes to `narrow_state.stream_sub`. I thought you were asking about this line specifically, which I changed to match the above.

Because `get_filtered_and_sorted_user_ids` returns `sort_users`, which was previously using `narrow_state.stream_sub` to get  the channel ID. But the test was only overriding `narrow_state.stream_id`, so the test previously wasn't doing any sorting of the buddy list users in that final step because both the channel ID and DM IDs were undefined/empty.

When I updated `stream_sub` to use `stream_id`, then the test failed until the current user (me) was properly added to the peer data for the channel.

evykassirer reviewed (commented):
Sorry for the delay! Left a couple thoughts :)

laurynmm commented:
**Updates**:
- Refactored and renamed `is_for_stream_id` to not get the `StreamSubscription` since we don't need to worry if the current channel ID is known or unknown.
- Refactored `get_first_unread_info` and the helper function it uses to have a defined `Filter` passed as a parameter since we expect this to be called with a message list.
- Updated `stream_id` to return `undefined` if we would have returned `NaN`, e.g., the operand is an empty string.
-  Added a boolean parameter to `stream_id` so that we can call it with that passed as `true` if we want to check our stream data for the ID that's in the operand.
- Final commit, which could be dropped as a follow up, looked as some places we can use that new functionality of `narrow_state.stream_id` to avoid return the `StreamSubscription` object when all we want is the ID.

laurynmm commented on web/src/buddy_data.ts:
```diff
@@ -363,7 +363,7 @@ function maybe_shrink_list(
     // We want to always show PM recipients even if they're inactive.
     const pm_ids_set = narrow_state.pm_ids_set();
-    const stream_id = narrow_state.stream_id();
+    const stream_id = narrow_state.stream_id(narrow_state.filter(), true);
```
We could also call this with `undefined` for the `Filter` parameter, but that seemed like it might be confusing when reading the code.

timabbott commented:
Merged, after adding some small comments.
